Ensure the daemon doesn't have fds open on the tty

krader1961 opened the pull request:
The Elvish daemon should not inherit file descriptors open on the tty.
Make the daemon's stdin open on the null device and its stdout/stderr
open on a predictable file.

Fixes #1191

xiaq commented on pkg/daemon/spawn.go:
```diff
@@ -58,7 +59,7 @@ func Spawn(cfg *SpawnConfig) error {
 	dbPath := abs("DbPath", cfg.DbPath)
 	sockPath := abs("SockPath", cfg.SockPath)
-	logPathPrefix := abs("LogPathPrefix", cfg.LogPathPrefix)
+	logPath := abs("LogPathPrefix", cfg.LogPathPrefix+strconv.Itoa(os.Getpid()))
```
Hmm, IIUC this PR moves the opening of the log file from the daemon process to its parent. Since the filename is now determined in the parent, it no longer reflects the PID of the daemon process itself, but that of the parent.

Why is this necessary, and is it related to the main purpose of this PR (ensure the daemon doesn't have fds open on the tty)?

krader1961 replied:
Yes, this does what you described. The reason is we don't want to risk losing any messages written to stdout/err before the daemon setup code can open a log file and redirect its os.Stdout/err to that log file. This is also marginally simpler than doing it in the daemon setup code. I'm also convinced that sooner, rather than later, the daemon will be eliminated. Too, this allows eliminating the `-logprefix` flag which shouldn't, and isn't mean to be, used by a person. So the fix employed by this P.R. makes more sense.

That the PID on the file doesn't match that of the daemon isn't really important. Note that the daemon explicitly records its PID in the first log message it writes. Regardless of which process names the file the PID could still match a prior execution and thus a log file that already exists.

xiaq replied:
> The reason is we don't want to risk losing any messages written to stdout/err before the daemon setup code can open a log file and redirect its os.Stdout/err to that log file.

That is not correct. The stdout and stderr of the daemon are not redirected to the log file specified by `-log` or `-logprefix`; those flags affect the destination of the "loggers" defined in https://github.com/elves/elvish/tree/master/pkg/logutil. There is actually a TODO saying that the stdout and stderr of the daemon probably should be redirected, though :): https://github.com/elves/elvish/blob/3615988408b7105c92096a784d915ca2568a83cb/pkg/daemon/spawn.go#L75

> That the PID on the file doesn't match that of the daemon isn't really important.

I agree and disagree. I agree that it's not important to have the PID as part of the filename since the first message is reporting the PID of the daemon. However, having another mysterious number that looks like a PID but is **not** the PID of the daemon can be very confusing.

It's better to just use a serial number; so instead of `daemon.log-23178` or `daemon.log-12865` you get `daemon.log-1`, `daemon.log-2`, etc.. I've actually implemented a function to support this idea, but so far the function is unused:

https://github.com/elves/elvish/blob/3615988408b7105c92096a784d915ca2568a83cb/pkg/fsutil/claim.go#L15

For the sake of this PR, I'd prefer that you leave this part out, since it's not necessary for the more important change of ensuring the daemon doesn't have FDs open on the TTY.

krader1961 replied:
> That is not correct. The stdout and stderr of the daemon are not redirected to the log file specified by -log or -logprefix

It is true with this change since I modified the code to pass os.Stdout to the logger. See the block at line 111 in *pkg/prog/prog.go*. I overlooked the TODO. That comment should also be removed.

krader1961 replied:
Hold on, I did remove the TODO about redirecting stdout/stderr for the daemon.

xiaq replied:
Please address this feedback:

> That the PID on the file doesn't match that of the daemon isn't really important.

I agree and disagree. I agree that it's not important to have the PID as part of the filename since the first message is reporting the PID of the daemon. However, having another mysterious number that looks like a PID but is **not** the PID of the daemon can be very confusing.

It's better to just use a serial number; so instead of `daemon.log-23178` or `daemon.log-12865` you get `daemon.log-1`, `daemon.log-2`, etc.. I've actually implemented a function to support this idea, but so far the function is unused:

https://github.com/elves/elvish/blob/3615988408b7105c92096a784d915ca2568a83cb/pkg/fsutil/claim.go#L15

krader1961 replied:
Okay, but you wrote in your previous comment immediately after talking about possibly using `ClaimFile`:

> For the sake of this PR, I'd prefer that you leave this part out, since it's not necessary for the more important change of ensuring the daemon doesn't have FDs open on the TTY.

Also, these are *temporary* files that are removed on each reboot and it should be rare for users to have more than one of these daemon log files. But since `ClaimFile` already exists I have no objection to using it. Updated PR in a few minutes....

krader1961 replied:
FYI, I found a bug in `ClaimFile`. It fails to create the file in the directory that it scanned for existing files that match the pattern. Instead it creates the file in the CWD. I've fixed that but didn't bother to update its unit test.

xiaq replied:
> Okay, but you wrote in your previous comment immediately after talking about possibly using ClaimFile:

Oh, I originally thought that this PR contains two independent changes, (1) making sure that the daemon doesn't have FDs on the TTY and (2) redirecting the stdout and stderr of the daemon, so I was asking you to make this PR only do (1) but not (2) - by "that" I meant part (2). But later I realized that those two things are the same thing.

xiaq commented on pkg/fsutil/claim.go:
```diff
@@ -53,7 +54,7 @@ func ClaimFile(dir, pattern string) (*os.File, error) {
 
 	for i := max + 1; ; i++ {
-		name := prefix + strconv.Itoa(i) + suffix
+		name := filepath.Join(dir, prefix+strconv.Itoa(i)+suffix)
```
Thanks for fixing the bug, but please do add a unit test.

krader1961 replied:
Isn't that your responsibility since I don't think modifying this change to use `ClaimFile` is even necessary or particularly useful? Note that I wouldn't have even noticed this problem if I didn't perform due diligence to see what happens if I launch an Elvish shell built with this change without an already accessible daemon.

xiaq replied:
Well, you could argue that everything is my responsibility and I wouldn't disagree :)

It's good practice to add a regression test when you fix a bug, but if doing that demotivates you from working on this PR I certainly won't insist.

xiaq commented on pkg/prog/prog.go:
```diff
@@ -55,7 +54,6 @@ func newFlagSet(stderr io.Writer, f *Flags) *flag.FlagSet {
 	fs.Usage = func() { usage(stderr, fs) }
 
 	fs.StringVar(&f.Log, "log", "", "a file to write debug log to")
```
Update the description for `-log` that it is ignored when `-daemon` is specified, since the daemon always writes logs to stdout.

krader1961 replied:
Done. However, I think this is a silly nitpick since manually launching the daemon is dangerous, not officially supported, and hopefully irrelevant in the future when the daemon is eliminated.

xiaq replied:
> I think this is a silly nitpick since manually launching the daemon is dangerous, not officially supported

Your premise is true from a user's perspective, but not true from a developer's perspective. You may not like the daemon, but as long as it exists as a developer I prefer making it easy to debug it and launching it manually is a valuable technique to do that.

xiaq commented on pkg/prog/prog.go:
```diff
@@ -110,13 +108,15 @@ func Run(fds [3]*os.File, args []string, programs ...Program) int {
 
-	if f.Log != "" {
+	if f.Daemon {
+		// See daemon.Spawn(). We expect our stdout file handle is open on a
+		// unique log file for the daemon to write its log messages.
```
I find it more readable to put references at the end; it'd also be better to include the package name too. So I'd write this as:

> We expect our stdout file handle is open on a unique log file for the daemon to write its log messages. See daemon.Spawn() in pkg/daemon.

krader1961 replied:
Okay. Done. However, I think this is nitpicking.

xiaq commented on pkg/shell/paths.go:
```diff
@@ -28,8 +27,8 @@ func MakePaths(stderr io.Writer, overrides Paths) Paths {
 	p := overrides
 	setDir(&p.RunDir, "secure run directory", getSecureRunDir, stderr)
 	if p.RunDir != "" {
+		setChild(&p.RunDir, p.RunDir, "")
```
Hmm?

xiaq replied:
Did you miss this comment? This line of code doesn't do anything useful and is confusing, although it doesn't seem to do any harm so I can fix this after merging this PR.

krader1961 replied:
Yes, that particular change is wrong. It's left over from earlier changes to this code when I was figuring out how to adapt the code to use `ClaimFile`.

xiaq reviewed (commented):
Overall LGTM now, some feedback for stylistic matters.

krader1961 commented on pkg/daemon/spawn.go:
```diff
@@ -18,13 +20,12 @@ type SpawnConfig struct {
 
 // Spawn spawns a daemon process in the background by invoking BinPath, passing
-// DbPath, SockPath and LogPathPrefix as command-line arguments after resolving
+// DbPath, SockPath and LogPath as command-line arguments after resolving
```
Done

krader1961 commented on pkg/daemon/spawn_unix.go:
```diff
@@ -7,11 +7,15 @@ import (
 
-func procAttrForSpawn() *os.ProcAttr {
+func procAttrForSpawn(stdout *os.File) *os.ProcAttr {
+	// The daemon should not inherit a reference to the stdin file descriptor
+	// of the original shell. The daemon shouldn't read anything from stdin so
+	// use the null device.
+	devnull, _ := os.OpenFile("/dev/null", os.O_RDONLY, 0)
```
**No. Don't ever do that for stdin/stdout/stderr.** I've debugged two vexing problems in the past due to programs that launched other programs with one or more of those file descriptors closed. UNIX programs read from, or write to, those file descriptors not just in the "main" program but also imported packages (libraries). Every program assumes they are open on something. If they aren't the first `open()` call the program executes will allocate one of those file descriptors. Oops! You now have other parts of the program that use the stdin/out/err file descriptors reading/writing a file not meant to be used as stdin/out/err.

It is extremely unlikely a different path will ever be needed for a Go module predicated on the meta GOOS value "unix".

xiaq replied:
That's a good point, but it seems that it applies more to stdout and stderr, which libraries are likely to write to. I'd be surprised if a library reads from stdin. Nonetheless, it sounds like good practice to avoid closing any of the lowest 3 FDs.

krader1961 commented on pkg/daemon/spawn_unix.go:
```diff
@@ -7,11 +7,15 @@ import (
+	devnull, _ := os.OpenFile("/dev/null", os.O_RDONLY, 0)
 	return &os.ProcAttr{
-		Dir:   "/",        // cd to /
+		Dir:   "/",        // make the CWD when the daemon starts the root directory
```
Fair enough, I have removed it; although I think it has some, slight, value to a reader of the code not intimately familiar with the semantics of `os.ProcAttr` and `os.StartProcess`.  Your observation also applies to the comment of the `Env` struct member so I've removed it as well.

krader1961 commented on pkg/daemon/spawn_windows.go:
```diff
@@ -15,11 +15,18 @@ const (
-		Dir:   `C:\`,
-		Env:   []string{"SystemRoot=" + os.Getenv("SystemRoot")}, // SystemRoot is needed for net.Listen for some reason
-		Files: []*os.File{os.Stdin, os.Stdout, os.Stderr},
+		Dir:   `C:\`, // make the CWD when the daemon starts the root directory
```
Done

krader1961 commented on pkg/daemon/spawn_windows.go:
```diff
@@ -15,11 +15,18 @@ const (
+	// of the original shell. The daemon shouldn't read anything from stdin so
+	// use the null device.
+	devnull, _ := os.OpenFile("NUL:", os.O_RDONLY, 0)
+	env := []string{
+		"SystemRoot=" + os.Getenv("SystemRoot"), // this is needed for net.Listen for some reason
+	}
```
I made that change because the literal is non-trivial (unlike in *spawn_unix.go*) and I was already touching the adjacent lines. But I don't feel strongly enough to argue for keeping this aspect of the P.R. so I've reverted it.

xiaq commented on pkg/daemon/spawn.go:
```diff
@@ -69,11 +70,17 @@ func Spawn(cfg *SpawnConfig) error {
+	}
+	defer daemonStdout.Close()
+
+	procattrs := procAttrForSpawn(daemonStdout)
+	_, err = os.StartProcess(binPath, args, procattrs)
+	procattrs.Files[0].Close() // drop our reference to the /dev/null handle passed to the daemon
```
Instead of letting `procAttrForSpawn` open `/dev/null` or `NUL` and relying on the implicit fact that `procattrs.Files[0]` is that file, it's better to open that file here and pass it as an argument to `procAttrForSpawn`; `os.DevNull` can be used as the filename.

xiaq reviewed (commented):
You sound frustrated, and I am sorry about that.

I appreciate the work you have put into PRs and I know it can be quite frustrating to wake up to a lot of comments and find that most of them are "nitpicking" ones.

But let me clarify my motivations too. Elvish is more or less my personal project, I take ownership over every line of code I merge, so I need make myself comfortable with every line of diff. I will continue point out all the places I am not comfortable with - and yes, I **am** picky when it comes to comments and documentation often in my own way.

**However**, I try to not block PRs purely on stylistic issues, since I can always easily make those changes in a followup commit. When I say *overall LGTM*, it means that only minor or stylistic issues remain, and it's OK for you to not act upon the remaining feedback, and leave it to me to carry out the changes; but I'll make comments nonetheless, as a note to myself.

I've done that proactively for some past PRs when review has dragged on for too long. For PRs where I don't do that proactively but you'd like me to, you can also explicitly say that you wish to have the PR merged and you'll leave those smaller changes to me.

Also, it seems that you have responded to most my comments but haven't pushed the code that addresses them? If you prefer to let me do the remaining stylistic changes, I can certainly do it in the PR's current form (I've already said *overall LGTM*), but if you already have a local commit it's better for you to push first.

xiaq commented:
In retrospect, when I said "overall LGTM", I should have explicitly offered you the option of having the PR merged as is, and made it your choice whether you want to further polish the PR or leave the polishing to me. I'll do that in future reviews.

krader1961 commented:
> You sound frustrated, and I am sorry about that.

I have no objection to being uncompromising (aka being a hard-ass) on proposed changes. I'm the same way when reviewing changes at $day-job. What pushed me over the edge was asking me to add (or fix) a unit test for `ClaimFile` since I

*) did not write that code,
*) spent an hour adapting my change to use `ClaimFile` despite my belief using it in the context of this change is pointless, and
*) the incorrect behavior of `ClaimFile` cost me another hour because I initially assumed my use of that function was incorrect, before I realized that function had a bug.

P.S., I have, partially, addressed your other point about using `os.DevNull` rather than hard-coded strings. I say "partially" because I have replaced the literal strings with that symbol but am not convinced moving those file opens to the caller is preferable. I want to sleep on that.

krader1961 commented:
@xiaq, This seems like a good place to comment on the fact that you seemingly never do a squash-merge of third-party changes. That caused me to switch from pushing deltas in response to your feedback and instead squashing fixups locally and force pushing so there is a single commit to be merged. I think pushing delta commits makes resolving feedback easier but perhaps you prefer what I've been doing, which is fine, but I have wondered if my local squash commits and forced pushes is counterproductive to getting changes merged sooner, rather than later.

xiaq commented:
> @xiaq, This seems like a good place to comment on the fact that you seemingly never do a squash-merge of third-party changes. That caused me to switch from pushing deltas in response to your feedback and instead squashing fixups locally and force pushing so there is a single commit to be merged. I think pushing delta commits makes resolving feedback easier but perhaps you prefer what I've been doing, which is fine, but I have wondered if my local squash commits and forced pushes is counterproductive to getting changes merged sooner, rather than later.

https://github.com/elves/elvish/pull/949#issuecomment-605340410

xiaq commented:
I guess this can be merged now.
